File: 110-BalancedBTree/BalancedBTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int getHeight(TreeNode* root){
        if(!root)   return 0;
        int l=getHeight(root->left);
        if(l==-1)   return -1;
        int r=getHeight(root->right);
        if(r==-1)   return -1;
        if(abs(l-r)>1) return -1;
        else return 1+max(l,r);
    }

    bool isBalanced(TreeNode* root) {
        if(!root)   return true;
        if(getHeight(root)==-1) return false;
        return true;
        
    }
};

File: 110-BalancedBTree/balancedBTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public boolean isBalanced(TreeNode root) {
        //height is counted by paths.
        if (root == null) return true;
        int temp = Math.abs(height(root.left) - height(root.right));
        return temp <=1 && isBalanced(root.left) && isBalanced(root.right);
        
        
    }
    
    public int height(TreeNode root) {
        if (root == null) return -1;
        int lh = height(root.left);
        int rh = height(root.right);
        return lh < rh ? rh + 1: lh +1;
        // int temp = heigth(root.left) < height(root.right) ? height(r)
    }
}